Documentation/coding-guidelines/interop-pinvokes.md

```diff
@@ -0,0 +1,239 @@
+
+**Blittable types:**
+
+- `byte`, `sbyte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `single`, `double`
+- non-nested one dimensional arrays of blittable types (e.g. `int[]`)
+- structs and classes with fixed layout that only have blittable types for instance fields
+  - fixed layout requires `[StructLayout(LayoutKind.Sequential)]` or `[StructLayout(LayoutKind.Explicit)]`
+  - structs are `LayoutKind.Sequential` by default, classes are `LayoutKind.Auto`
+
+**NOT blittable:**
+
+- `bool`
+
+**SOMETIMES blittable:**
+
+- `char`, `string`
+
+When blittable types are passed by reference they are simply pinned by the marshaller instead of being copied to an intermediate buffer. (Classes are inherently passed by reference, structs are passed by reference when used with `ref` or `out`.)
+
+`char` is blittable in a one dimensional array **or** if it is part of a type that contains it is explicitly marked with `[StructLayout]` with `CharSet = CharSet.Unicode`.
+
+```C#
+[StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
+public struct UnicodeCharStruct
+{
+    public char c;
+}
+```
+
+`string` is blittable if it isn't contained in another type and it's being passed as an argument that is marked with `[MarshalAs(UnmanagedType.LPWStr)]` or the `[DllImport]` has `CharSet = CharSet.Unicode` set.
+
+You can see if a type is blittable by attempting to create a pinned `GCHandle`. If the type is not a string or considered blittable `GCHandle.Alloc` will throw an `ArgumentException`.
+
+
+[Blittable and Non-Blittable Types](https://msdn.microsoft.com/en-us/library/75dwhxf7.aspx "MSDN")  
+[Default Marshalling for Value Types](https://msdn.microsoft.com/en-us/library/0t2cwe11.aspx "MSDN")
+
+Keeping Managed Objects Alive
+-----------------------------
+`GC.KeepAlive()` will ensure an object stays in scope until the KeepAlive method is hit.
+
+[`HandleRef`][6] allows the marshaller to keep an object alive for the duration of a P/Invoke. It can be used instead of `IntPtr` in method signatures. `SafeHandle` effectively replaces this class and should be used instead.
+
+[6]: https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.handleref.aspx "MSDN"
+
+[`GCHandle`][7] allows pinning a managed object and getting the native pointer to it. Basic pattern is:  
+
+``` C#
+GCHandle handle = GCHandle.Alloc(obj, GCHandleType.Pinned);
+IntPtr ptr = handle.AddrOfPinnedObject();
+handle.Free();
+```
+
+[7]: https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.gchandle.aspx "MSDN"
+
+Pinning is not the default for `GCHandle`. The other major pattern is for passing a reference to a managed object through native code back to managed code (via a callback, typically). Here is the pattern:
+
+``` C#
+GCHandle handle = GCHandle.Alloc(obj);
+SomeNativeEnumerator(callbackDelegate, GCHandle.ToIntPtr(handle));
+
+// In the callback
+GCHandle handle = GCHandle.FromIntPtr(param);
+object managedObject = handle.Target;
+```
```

Comment:
You should also say here that the GCHandle needs to be explicitly freed to avoid memory leaks.